libbeat/processors/actions/truncate_fields.go

```diff
@@ -0,0 +1,180 @@
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package actions
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+type truncateFieldsConfig struct {
+	Fields        []string `config:"fields"`
+	MaxBytes      int      `config:"max_bytes" validate:"min=0"`
+	MaxCount      int      `config:"max_character_count" validate:"min=0"`
+	IgnoreMissing bool     `config:"ignore_missing"`
+	FailOnError   bool     `config:"fail_on_error"`
+}
+
+type truncateFields struct {
+	config   truncateFieldsConfig
+	truncate truncater
+}
+
+type truncater func(*truncateFields, []byte) ([]byte, error)
+
+func init() {
+	processors.RegisterPlugin("truncate_fields",
+		configChecked(newTruncateFields,
+			requireFields("fields"),
+			mutuallyExclusiveRequiredFields("max_bytes", "max_count"),
+		),
+	)
+}
+
+func newTruncateFields(c *common.Config) (processors.Processor, error) {
+	var config truncateFieldsConfig
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("fail to unpack the truncate_fields configuration: %s", err)
+	}
+
+	var truncateFunc truncater
+	if config.MaxBytes > 0 {
+		truncateFunc = (*truncateFields).truncateBytes
+	} else {
+		truncateFunc = (*truncateFields).truncateCharacters
+	}
+
+	return &truncateFields{
+		config:   config,
+		truncate: truncateFunc,
+	}, nil
+}
+
+func (f *truncateFields) Run(event *beat.Event) (*beat.Event, error) {
+	var backup common.MapStr
+	if f.config.FailOnError {
+		backup = event.Fields.Clone()
+	}
+
+	for _, field := range f.config.Fields {
+		event, err := f.truncateSingleField(field, event)
+		if err != nil && f.config.FailOnError {
+			logp.Debug("truncate_fields", "Failed to truncate fields: %s", err)
+			event.Fields = backup
+			return event, err
+		}
+	}
+
+	return event, nil
+}
+
+func (f *truncateFields) truncateSingleField(field string, event *beat.Event) (*beat.Event, error) {
+	v, err := event.GetValue(field)
+	if err != nil {
+		if f.config.IgnoreMissing && errors.Cause(err) == common.ErrKeyNotFound {
+			return event, nil
+		}
+		return event, fmt.Errorf("could not fetch value for key: %s, Error: %+v", field, err)
```

Comment:
Not sure if this logging format with `"Error: %+v"` is an existing convention in this package, but if not I'd recommend using `errors.Wrapf` to get more consistent error formatting with other parts of the codebase. You'd just write `errors.Wrapf("could not fetch value for key: %s", field)`.